data.md

```diff
@@ -955,5 +1007,154 @@ Decoding
     rule #decodeLengthPrefixLength(#str,  STR, START, B0) => #decodeLengthPrefixLength(#str,  START, B0 -Int 128 -Int 56 +Int 1, #asWord(#parseByteStackRaw(substrString(STR, START +Int 1, START +Int 1 +Int (B0 -Int 128 -Int 56 +Int 1)))))
     rule #decodeLengthPrefixLength(#list, STR, START, B0) => #decodeLengthPrefixLength(#list, START, B0 -Int 192 -Int 56 +Int 1, #asWord(#parseByteStackRaw(substrString(STR, START +Int 1, START +Int 1 +Int (B0 -Int 192 -Int 56 +Int 1)))))
     rule #decodeLengthPrefixLength(TYPE, START, LL, L) => TYPE(L, START +Int 1 +Int LL)
+```
+
+Merkle Patricia Tree
+====================
+
+- Appendix C and D from the Ethereum Yellow Paper
+- https://github.com/ethereum/wiki/wiki/Patricia-Tree
+
+```k
+    syntax KItem ::= Int | MerkleTree // For testing purposes
+
+    syntax MerkleTree ::= MerkleBranch    ( Map, String )
+                        | MerkleExtension ( ByteArray, MerkleTree )
+                        | MerkleLeaf      ( ByteArray, String )
+                        | ".MerkleTree"
+                        | ".MerkleBranch"            [function]
+ // -----------------------------------------------------------
+    rule .MerkleBranch
+      => MerkleBranch (  0 |-> .MerkleTree  1 |-> .MerkleTree  2 |-> .MerkleTree  3 |-> .MerkleTree
+                         4 |-> .MerkleTree  5 |-> .MerkleTree  6 |-> .MerkleTree  7 |-> .MerkleTree
+                         8 |-> .MerkleTree  9 |-> .MerkleTree 10 |-> .MerkleTree 11 |-> .MerkleTree
+                        12 |-> .MerkleTree 13 |-> .MerkleTree 14 |-> .MerkleTree 15 |-> .MerkleTree
+                      , ""
+                      )
```

Comment:
It looks like MerkleBranch always has exactly 16 entries? I suspect that for real-world executions we'll see a pretty large slowdown because of the `log(N)` map accesses/updates that will pile up when doing this. Some alternatives I can think of are (1) using a constructor with 16 arguments directly, or (2) using a `List` (but it looks like there are places where you aren't only appending to beginning/end below).

One thought might be to try record syntax:

```
syntax MerkleTree ::= MerkleBranch ( MerkleChildren , String )
syntax MerkleChildren ::= MerkleChildren ( mt0: MerkleTree, mt1: MerkleTree , mt2: MerkleTree , mt3: MerkleTree , ... , mt15: MerkleTree )
```

Not sure if that would be helpful, but you don't need to use the record syntax. But it looks like the only place you use the fact that it's a `Map` is when you get/set at specific `PATH` in the subtrees. So I think you can write those getters/setters directly over `MerkleChildren` instead (make some new syntax for it).

For example:

```
syntax MerkleTree ::= #getMerklePath ( MerkleTree , Int ) [function]
rule #getMerklePath ( MerkleChildren ( MT , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ ) , 0 ) => MT
rule #getMerklePath ( MerkleChildren ( _ , MT , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ ) , 1 ) => MT
...
rule #getMerklePath ( MerkleChildren ( _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , _ , MT ) , 15 ) => MT
```

Similarly you can write a `#getMerklePath` as well.

Main benefit will be performance of concrete execution and pattern matching.

@dwightguth any thoughts? Do you think this will have significant impact for large merkle trees?